### DISCUSSION TITLE
@Transactional의 readOnly를 true로 설정했을 때 이점은 무엇일까요?

### DISCUSSION THREAD

**Discussion by Hanjaemo** [Hanjaemo/DB-Study#15] [Q&A]
Spring에서 제공하는 트랜잭션 관리 어노테이션 `@Transactional`에는 readOnly라는 옵션이 있습니다.
어떤 트랜잭션이 조회만 수행한다면 readOnly를 true로 설정하여 성능상 이점을 챙기라고 합니다.
이때 어떤 부분에서 성능상 이점이 존재하는지 궁금합니다.

**Comment by mithzinf** [comment]
@Transactional(readOnly=true) 옵션을 설정하면 해당 트랜잭션은 CRUD 중 Read만 실행한다는 의미로, 데이터를 조회만 하고 변경하지 않다는 것을 약속하게 됩니다.
이 옵션을 쓰게 되면 생기는 성능 면에서 여러가지 이점은 다음와 같다고 생각합니다.

1. 성능 향상 
읽기 외 DB에 관한 INSERT 작업을 하지 않기 때문에, DB에 불필요한 로그가 감소됩니다. 그리고 데이터가 변경되지 않기 때문에, 데이터 변경을 위한 락 기능을 사용하지 않아도 되기 때문에 락 관리를 최적화할 수 있습니다.

2. 더티 체킹 불필요
더티 체킹 : 엔티티 상태 변화를 자동 감지 & 트랜잭션이 커밋되는 시점에 변경된 사항을 DB에 반영하는 영속성 컨텍스트에 의해 관리되고 실행되는 기능
데이터를 조회만 하고 변경은 하지 않는 트랜잭션 옵션이므로, 자연스레 더티 체킹 생략함으로써 성능 향상 & 리소스 절약

**Comment by gomudayya** [comment]
주된 이유로는 JAP를 사용할 때, 더티체킹(변경감지)에 대한 비용을 없앤다는 것 같습니다.

관련된 JPA코드인데 주석에 설명이 잘 되어있네요



```java
if ( isReallyReadOnly ) {
    //no need to take a snapshot - this is a                   
    //performance optimization, but not really
    //important, except for entities with huge
    //mutable property 
    persistenceContext.setEntryStatus( entityEntry, Status.READ_ONLY );
}
else {
    //take a snapshot
    TypeHelper.deepCopy(
            hydratedState,
            persister.getPropertyTypes(),
            persister.getPropertyUpdateability(),
            //after setting values to object
            hydratedState,
            session
    );
    persistenceContext.setEntryStatus( entityEntry, Status.MANAGED );
}
```
[주석 내용]

readOnly일 경우에는 스냅샷이 찍을필요가 없고, 이렇게 함으로써 성능최적화가 된다.
그러나 정말 대규모의 엔티티를 받아와서 변경하는 경우를 제외하곤 그다지 중요하진 않다

라고 합니다.

[참고](https://perfectacle.github.io/2021/08/08/readonly-transaction-doesnt-make-entity-snapshot/)

**Comment by jschoi-96** [comment]
JPA의 더티체킹이란, 트랜잭션이 끝나는 시점에 변화가 있는 모든 엔티티 객체를 데이터베이스에 자동으로 반영해줍니다. 

JPA는 스냅샷을 통해서 엔티티의 조회 상태를 찍어놓고, 변경 사항이 있다면 update 쿼리를 데이터베이스에 전달합니다.

하지만, `@Transactional(readOnly = true)`를 사용한다는 뜻은 트랜잭션을 읽기 전용으로 설정한다는 뜻이고 이에 따라서 더티 체킹을 생략함으로써 메모리를 절약하고 데이터베이스에 쓰기 작업을 수행하지 않게 됩니다. 

참고자료: https://studyandwrite.tistory.com/573
